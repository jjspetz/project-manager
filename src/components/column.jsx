import React, { Component } from 'react';
// Material UI imports
import {Card, CardTitle} from 'material-ui/Card';
import FloatingActionButton from 'material-ui/FloatingActionButton';
import ContentAdd from 'material-ui/svg-icons/content/add';
import Task from './task';
import {connect} from 'react-redux';
import {toggleInput} from '../actions';
import {bindActionCreators} from 'redux';
import database, {User} from '../fire.js'
// react dnd
import PropTypes from 'prop-types';
import { ItemTypes } from '../constants/constants';
import { DropTarget } from 'react-dnd';

const dropTask = {
  drop(props, monitor, connect) {
    // rewrites db entry to now render in the new container
    database.ref('users/' + User.user.uid).child(monitor.getItem().id).set({
      column: connect.props.title,
      task: monitor.getItem().id
    });
  }
};

function collect(connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver()
  };
}

class Column extends Component {
  constructor(props) {
    super(props);
    this.state = {
      val: ''
    };
  }

  toggleInput = () => {
    this.props.toggleInput(this.props.openInput === this.props.title ? '' : this.props.title);
  }

  addTask = (event) => {
    // writes in database
    database.ref('users/' + User.user.uid).child(this.state.val).set({
      column: this.props.title,
      task: this.state.val
    });
    // closes input
    this.props.toggleInput('');
    // resets input to blank string
    this.setState({val: ''});
    event.preventDefault();
  }

  add(event, key) {
      this.setState({[key]: event.target.value})
    }

  render() {
    const { connectDropTarget} = this.props;
    return connectDropTarget(
      <div className='card'>
      <Card className='card'>
        <CardTitle title={this.props.title}/>
        <ol id={this.props.title}>
        {this.props.tasks ? Object.values(this.props.tasks).map((task) =>
          task.column === this.props.title ?
          <Task key={task.task} task={task}/>
          : null) : null}
        </ol>
        <FloatingActionButton onTouchTap={this.toggleInput} mini={true} className='add'>
          <ContentAdd />
        </FloatingActionButton>

        {this.props.openInput === this.props.title ?
        <form className='input' onSubmit={event => this.addTask(event)}>
          <input type='text'
            value={this.state.val}
            onChange={event => this.add(event, 'val')} />
          <input type='submit' value='Submit'/>
        </form>
        : null}

      </Card>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return {
    tasks: state.tasks,
    openInput: state.openInput,
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      toggleInput: toggleInput
    }, dispatch
  )
}

Column.propTypes = {
  isOver: PropTypes.bool.isRequired,
  place: PropTypes.number
};

Column = DropTarget(ItemTypes.TASK, dropTask, collect)(Column)
export default connect(mapStateToProps, mapDispatchToProps)(Column)
