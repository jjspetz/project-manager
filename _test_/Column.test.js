import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Column } from '../src/components/column.jsx';

const faux = el => el;
const wrapper = shallow(<Column connectDropTarget={faux} />);


describe('Column component', () => {
  it('should render', () => {
    expect(wrapper).toBeTruthy();
  });
  it('should match snapshot', () => {
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});
