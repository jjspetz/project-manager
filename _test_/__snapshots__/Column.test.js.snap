// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column component should match snapshot 1`] = `
<div
  className="flex-container"
>
  <Card
    className="card"
    expandable={false}
    expanded={null}
    initiallyExpanded={false}
  >
    <CardTitle
      titleStyle={
        Object {
          "fontSize": undefined,
        }
      }
    />
    <ol />
    <div
      className="lowerCard"
    >
      <FloatingActionButton
        className="add"
        disabled={false}
        mini={true}
        onTouchTap={[Function]}
        secondary={false}
        zDepth={2}
      >
        <ContentAdd />
      </FloatingActionButton>
      <form
        className="cardInput"
        onSubmit={[Function]}
      >
        <TextField
          disabled={false}
          floatingLabelFixed={false}
          fullWidth={false}
          hintText="Enter a task"
          multiLine={false}
          onChange={[Function]}
          rows={1}
          type="text"
          underlineShow={true}
          value=""
        />
        <RaisedButton
          buttonStyle={Object {}}
          disabled={false}
          fullWidth={false}
          label="Submit"
          labelPosition="after"
          labelStyle={Object {}}
          onTouchTap={[Function]}
          primary={true}
          secondary={false}
          style={Object {}}
        />
      </form>
    </div>
  </Card>
</div>
`;
